File: src/main/java/me/giocode/FormatNumber.java
package me.giocode;

public class FormatNumber {


    public static String solution(String input) {

        String cleanedInput = input.replace(" ", "").replace("-", "");
        StringBuilder result = new StringBuilder();
        int currentIndex = 0;
        int len = cleanedInput.length();
        int remaining;
        int increment;

        while (currentIndex < len) {
            remaining = len - currentIndex;

            if (currentIndex > 0) {
                result.append("-");
            }

            if (remaining > 4 || remaining == 3) {
                increment = 3;
            } else {
                increment = 2;
            }
            result.append(cleanedInput.substring(currentIndex, currentIndex + increment));
            currentIndex += increment;

        }

        return result.toString();
    }
}

File: src/main/kotlin/leetcode/FormatNumber.kt
package leetcode

//https://leetcode.com/discuss/post/1155893/toptal-oa-format-number-by-eels-dev-diav/

class FormatNumber {

    fun solution(input: String): String {
        val cleanedInput = input.replace(" ", "").replace("-", "")
        val result = StringBuilder()
        var currentIndex = 0

        while (currentIndex < cleanedInput.length) {
            val remaining = cleanedInput.length - currentIndex
            val groupSize = if (remaining > 4 || remaining == 3) 3 else 2

            if (result.isNotEmpty()) {
                result.append("-")
            }

            result.append(cleanedInput, currentIndex, currentIndex + groupSize)
            currentIndex += groupSize
        }

        return result.toString()
    }
}
